feat(cli): support XDG base directory specification

bacongobbler opened the pull request:
This PR makes Helm store configuration by default in `$XDG_CONFIG_HOME/helm`.

originated from #5166
closes #4491
closes #5386

Commit:
feat(cli): support XDG base directory specification

Signed-off-by: Matthew Fisher <matt.fisher@microsoft.com>

bacongobbler commented:
marking as a WIP so I can give this a quick spin

ian-howell commented on pkg/cli/environment.go:
```diff
@@ -26,14 +26,21 @@ import (
+//  - windows: %APPDATA%\helm
+//  - darwin:  ~/Library/Preferences/helm
+//  - linux:   ~/.config/helm
+var defaultHelmHome = filepath.Join(xdg.ConfigHome(), "helm")
```
This has the same problem as what was discussed in https://github.com/helm/helm/pull/5166#pullrequestreview-194324802. It's resolved by using
```go
var defaultHelmHome = xdg.App{}.ConfigPath("helm")
```

bacongobbler replied:
Thanks for catching that!

ian-howell reviewed (changes requested):
Throughout this change, you're referring to an environment variable called XDG_CONFIG_DIR. This variable is not a part of the [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html). The variable you should be using is XDG_CONFIG_HOME

bacongobbler commented:
I decided to vendor in xdg-go as there was a lack of unit tests, some shifty design questions around `defaultApp`, and how the package interacts with `os.Getenv("HOME")` on *nix distributions... $HOME is not set for some OS distributions.

Still ironing out a few kinks but should have ti ready by this afternoon.

bacongobbler commented:
Okay, this seems to be ready for review.

```
><> ./bin/helm init
Creating /home/bacongobbler/.config/helm 
Creating /home/bacongobbler/.config/helm/repository 
Creating /home/bacongobbler/.config/helm/repository/cache 
Creating /home/bacongobbler/.config/helm/plugins 
Creating /home/bacongobbler/.config/helm/starters 
Creating /home/bacongobbler/.config/helm/cache/archive 
Creating /home/bacongobbler/.config/helm/repository/repositories.yaml 
Adding stable repo with URL: https://kubernetes-charts.storage.googleapis.com 
$HELM_HOME has been configured at /home/bacongobbler/.config/helm.
Happy Helming!
><> HELM_HOME=/tmp/helm ./bin/helm init
Creating /tmp/helm 
Creating /tmp/helm/repository 
Creating /tmp/helm/repository/cache 
Creating /tmp/helm/plugins 
Creating /tmp/helm/starters 
Creating /tmp/helm/cache/archive 
Creating /tmp/helm/repository/repositories.yaml 
Adding stable repo with URL: https://kubernetes-charts.storage.googleapis.com 
$HELM_HOME has been configured at /tmp/helm.
Happy Helming!
```

ian-howell commented:
Is the intention to completely remove compatibility with $HOME/.helm?
If so, lgtm!

bacongobbler commented:
Yes. Backwards compatibility with the old ~/.helm was never intended as part of the original proposal. 3.0 is an opportunity to make these backwards-incompatible breaking changes, so we don't need to be legacy-compatible.

If users still want to use ~/.helm, they can set $HELM_HOME to ~/.helm if they so desire. However, they shouldn't expect that they can re-use their configuration from Helm 2 in Helm 3, as `helm home`'s layout is a massive departure from Helm 2. The two aren't compatible.

adamreese commented on pkg/xdg/xdg_test_windows.go:
```diff
@@ -0,0 +1,164 @@
+	os.Unsetenv("XDG_DATA_HOME")
+	os.Setenv("APPDATA", filepath.Join(homedir.HomeDir(), "foo"))
+
+	expected := filepath.Join(homedir.HomeDir(), "foo" appName, testFile)
```
missing some commas `"foo" appName` `"foo", appName`

bacongobbler replied:
good catch, fixed

bacongobbler commented on pkg/xdg/xdg_test_windows.go:
```diff
@@ -0,0 +1,164 @@
+)
+
+const (
+	appName  string = "helm"
```
anyone know how these test variables could be shared across windows/mac/unix builds so it's all DRY?

bacongobbler commented:
After pairing with @adamreese on this one, there's a lot going on here that we can remove. We don't need to implement the entire XDG spec for Helm's use case. Additionally we can clean up some of the code here if we made the package stateless. By doing so we could replace `pkg/helmpath` entirely with this (and potentially resolve issues like #5386 along the way). I'll look into this more. For now I'll put this back into the think tank.

bacongobbler commented:
self-reminder: remove all references to $HELM_HOME

ian-howell commented:
Just found this and thought it might be worth mentioning here: Go 1.13, which is expected to drop this month, will include a function [`os.UserConfigDir`](https://tip.golang.org/pkg/os/#UserConfigDir), which returns the same thing as helmpath.ConfigPath. The combination of this new function and the existing [`os.UserCacheDir`](https://golang.org/pkg/os/#UserCacheDir) will make most of `pkg/helmpath` redundant. I also found [this proposal](https://github.com/golang/go/issues/29960) which provides some background on the still missing `os.UserDataDir` function.

bacongobbler commented:
thanks @ian-howell! I'll go take a look over there and see where we can reuse the stdlib :)

thomastaylor312 reviewed (changes requested):
A few nits, a few questions, and a few code structure changes, but the functionality looks great!

bacongobbler commented:
Thanks @thomastaylor312! I'l get to addressing your comments tomorrow.

bacongobbler commented on cmd/helm/home.go:
```diff
@@ -23,28 +23,33 @@ import (
 
 func newHomeCmd(out io.Writer) *cobra.Command {
 	cmd := &cobra.Command{
 		Use:   "home",
```
Good point. Any suggestions on a new name? We probably want to add `home` as an alias for the new name either way. Some people will want to store data in `$XDG_CONFIG_DIR/helm`

bacongobbler replied:
I've decided to re-name it `helm path`, accepting optional arguments to display a specific directory.

bacongobbler commented on cmd/helm/home.go:
```diff
@@ -23,28 +23,33 @@ import (
 		Run: func(cmd *cobra.Command, args []string) {
-			h := settings.Home
-			fmt.Fprintln(out, h)
+			// NOTE(bacongobbler): the order here is important: we want to display the config path
```
SGTM

bacongobbler replied:
I've decided to re-name it `helm path`, accepting optional arguments to display a specific directory.

bacongobbler commented on cmd/helm/repo_remove.go:
```diff
@@ -54,11 +52,11 @@ func newRepoRemoveCmd(out io.Writer) *cobra.Command {
 }
 
 func (r *repoRemoveOptions) run(out io.Writer) error {
```
much better.

bacongobbler commented on cmd/helm/root.go:
```diff
@@ -100,7 +101,9 @@ Common actions from this point include:
 
 Environment:
-  $HELM_HOME          set an alternative location for Helm files. By default, these are stored in ~/.helm
+  $XDG_CACHE_HOME     set an alternative location for storing cached files.
```
These variables should use the XDG prefix as changing this variable affects the base directory. It doesn't change the `helm` directory itself, so this is correct.

As an example, by default on linux, this would be `~/.config/helm`. By changing $XDG_CACHE_HOME to `/tmp`, the alternative location is now `/tmp/helm`.

For example, on my Ubuntu machine, the Operating System sets $XDG_CACHE_HOME to `/home/bacongobbler/snap/code/common/.cache/` by default. So, the expected output of `helm path cache` is

```
><> ./bin/helm path cache
/home/bacongobbler/snap/code/common/.cache/helm
```

Does that make sense?

thomastaylor312 replied:
Yep, that makes sense

bacongobbler commented on cmd/helm/testdata/helmhome/helm/plugins/fullenv/fullenv.sh:
```diff
@@ -2,8 +2,9 @@
-echo $HELM_PATH_REPOSITORY_FILE
 echo $HELM_PATH_CACHE
+echo $HELM_PATH_CONFIG
+echo $HELM_PATH_DATA
```
Some systems infer the XDG base directory on system information, so plugins can't rely on the presence of XDG_CONFIG_HOME being present. So, Helm populates a set of environment variables to inform the plugin where it stores information to make it easier on plugin developers to infer where they can read that data.

bacongobbler commented on internal/test/ensure/ensure.go:
```diff
@@ -0,0 +1,65 @@
+}
+
+// TestHome creates a home directory like ensureHome, but without remote references.
+func TestHome(t *testing.T) {
```
since the package is already called `ensure`, I renamed it to `ensure.HomeDirs`.

bacongobbler commented on pkg/helmpath/xdg_darwin_test.go:
```diff
@@ -0,0 +1,85 @@
+var testApp = app{name: appName}
+
+func TestDataPath(t *testing.T) {
+	os.Unsetenv("XDG_DATA_HOME")
```
Yes, in case the user has this environment variable already set. Unsetting the environment variable ensures this will work on anyone's machine. We want to test the scenario where when the environment variable is unset, we use the default home path.

The `UnsetEnv` wasn't necessary for root_test.go, so I removed that. Good catch.

bacongobbler commented on pkg/repo/repotest/server.go:
```diff
@@ -35,22 +35,21 @@ import (
 //
```
Good point. I figure this package may be useful on its own for others who want to mock a chart repository, though. Most things in `internal` are all packages we definitely don't want SDK consumers to use (like `internal/version` or `internal/test`, which both rely heavily on Helm's internal architecture to work properly). It could be argued that `repotest` could be useful in certain scenarios like the charts-testing project.

Thoughts?

thomastaylor312 replied:
I did not consider the external testing needs for this one, so let's keep it here

bacongobbler commented on pkg/getter/getter_test.go:
```diff
@@ -50,13 +52,9 @@ func TestProviders(t *testing.T) {
+	os.Setenv("XDG_DATA_HOME", "testdata")
 
-	all := All(env)
+	all := All(cli.EnvSettings{})
```
I'm not sure how we could remove the struct and simplify the function signatures. Would you mind elaborating a little more on what you think we should do here?

thomastaylor312 replied:
Oh, now that I am following through the code again, I realized this isn't a dead struct. It looked similar to the other dead structs we had found

Commit:
fix(plugin): add HELM_HOME back

This allows Helm 2 plugins that used HELM_HOME as a scratchpad to continue to work the same in Helm 3.

Signed-off-by: Matthew Fisher <matt.fisher@microsoft.com>

hickeyma commented:
Starting using this. Like the standard feel and approach to it. Works well with plugin installs. Nice work @bacongobbler.  :clap:

One small thing. Should there be a corresponding doc PR for this?

bacongobbler commented:
Yep, working on it as well as the 3 way merge docs
